### PR TITLE
[SPARK-40222][SQL] Numeric try_add/try_divide/try_subtract/try_multiply should throw error from their children

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Similar to https://issues.apache.org/jira/browse/SPARK-40054, we should refactor the try_add/try_subtract/try_multiply/try_divide functions so that the errors from their children will be shown instead of ignored.
 Spark SQL allows arithmetic operations between Number/Date/Timestamp/CalendarInterval/AnsiInterval (see the rule [ResolveBinaryArithmetic](https://github.com/databricks/runtime/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L501) for details). Some of these combinations can throw exceptions too: * Date + CalendarInterval

- Date + CalendarInterval
- Date + AnsiInterval
- Timestamp + AnsiInterval
- Date - CalendarInterval
- Date - AnsiInterval
- Timestamp - AnsiInterval
- Number * CalendarInterval
- Number * AnsiInterval
- CalendarInterval / Number
- AnsiInterval / Number

This Jira is for the cases when both input data types are numbers. I will open jira tickets for DateTime types arithmetic operations if this one got merged.

### Why are the changes needed?

Fix the semantics of try_add/try_divide/try_subtract/try_multiply. 

### Does this PR introduce _any_ user-facing change?

Yes, after changes, the error from the children of try_add/try_divide/try_subtract/try_multiply functions will be shown instead of ignored.

### How was this patch tested?

Existing UT + new UT

### CODE REVIEW

**Comment by gengliangwang** [comment]
cc @srielau as well

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 1]
```diff
@@ -604,7 +604,9 @@ trait SupportQueryContext extends Expression with Serializable {
 
   def initQueryContext(): Option[SQLQueryContext]
 
-  def getContextOrNull(): SQLQueryContext = queryContext.getOrElse(null)
+  def getContextOrNull(): SQLQueryContext = {
+    queryContext.getOrElse(null)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:608]: unnecessary change?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TryEval.scala [hunk 2]
```diff
@@ -77,8 +77,13 @@ case class TryEval(child: Expression) extends UnaryExpression with NullIntoleran
 // scalastyle:on line.size.limit
 case class TryAdd(left: Expression, right: Expression, replacement: Expression)
     extends RuntimeReplaceable with InheritAnalysisRules {
-  def this(left: Expression, right: Expression) =
-    this(left, right, TryEval(Add(left, right, failOnError = true)))
+  def this(left: Expression, right: Expression) = this(left, right,
+    (left.dataType, right.dataType) match {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TryEval.scala:81]: hmm, are we sure `left` and `right` are resolved at this point?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala [hunk 3]
```diff
@@ -698,9 +751,14 @@ trait DivModLike extends BinaryArithmetic {
...
-    failOnError: Boolean = SQLConf.get.ansiEnabled) extends DivModLike {
+    evalMode: EvalMode.Value = EvalMode.fromSQLConf(SQLConf.get)) extends DivModLike {
+
+  def this(left: Expression, right: Expression) =
+    this(left, right, EvalMode.fromSQLConf(SQLConf.get))
 
-  def this(left: Expression, right: Expression) = this(left, right, SQLConf.get.ansiEnabled)
+  // `try_divide` has exactly the same behavior as the legacy divide, so here it only executes
+  // the error code path when `evalMode` is `ANSI`.
+  protected override def failOnError: Boolean = evalMode == EvalMode.ANSI
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala:761]: can we say that the `evalMode` will never be `TRY` in `Divide`? If so can we add an assert?

**Review by cloud-fan — APPROVED** [review]
LGTM in general. I think we need to skip handling datetime math expressions in `ResolveBinaryArithmetic`, where we know the data type of the inputs.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala (hunk 3 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala:761]: Yes, sure

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TryEval.scala (hunk 2 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TryEval.scala:81]: From tests, yes.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala (hunk 3 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala:761]: Yes and No. 
In the future we may support TRY eval mode in `Interval / Number`. Thus we need to keep the `TRY` eval mode and resolve it in https://github.com/apache/spark/pull/37663/files#diff-ed19f376a63eba52eea59ca71f3355d4495fad4fad4db9a3324aade0d4986a47R443
It doesn't seem quite necessary to add the assert now.

**Comment by gengliangwang** [comment]
@cloud-fan Thanks for the review.
Merging to master
